### ISSUE TITLE
What is exactly SoftmaxOutput do when backward?

### ISSUE THREAD

**Issue by BebDong** [dmlc/gluon-cv#1668]
I coincidentally found that doing any operation after the `SoftmaxOutput` operator yields no influence on the gradients when back-propagation. The [official docs](https://mxnet.apache.org/versions/1.8.0/api/python/docs/api/ndarray/ndarray.html?highlight=softmaxoutput#mxnet.ndarray.SoftmaxOutput) about `SoftmaxOutput` are quite confusing. With the following codes, I visualized the gradient histogram of trainable parameters.

```python
from mxnet import nd as F
from mxnet import gpu, autograd, random
from mxboard import SummaryWriter
from gluoncv.loss import MixSoftmaxCrossEntropyLoss
from gluoncv.model_zoo import get_fcn_resnet50_voc

# data
random.seed(128, ctx=gpu())
img = F.random.uniform(shape=(1, 3, 480, 480), ctx=gpu())
label = F.random.randint(low=0, high=21, shape=(1, 480, 480), ctx=gpu())

# model
model = get_fcn_resnet50_voc()
model.initialize(force_reinit=True)
model.collect_params().reset_ctx(gpu())

# loss
losses = [
    MixSoftmaxCrossEntropyLoss(aux=True, aux_weight=.1),
    MixSoftmaxCrossEntropyLoss(aux=True, aux_weight=.5),
    MixSoftmaxCrossEntropyLoss(aux=True, aux_weight=.9),
    MixSoftmaxCrossEntropyLoss(aux=True, aux_weight=5.)
]

# sw
sw = SummaryWriter(logdir='../logs')

# auto-grad
for i, criterion in enumerate(losses):
    criterion.initialize(ctx=gpu())
    with autograd.record():
        loss = criterion(*model(img), label)
    autograd.backward(loss)
    print(loss)
    param_names = [j.name for j in model.collect_params().values() if j.grad_req != 'null']
    param_grads = [j.grad() for j in model.collect_params().values() if j.grad_req != 'null']
    for k, name in enumerate(param_names):
        sw.add_histogram(tag=name, values=param_grads[k], global_step=i, bins=1000)

sw.close()
```

I then found no significant difference in gradients of `conv` layers in the histograms where the right axis indicates `MixSoftmaxCrossEntropyLoss` of different `aux_weight`. Minor differences are seemingly caused by data precision. For example, 

![image](https://user-images.githubusercontent.com/22273954/120605771-0940d080-c481-11eb-9364-2a3b590837ae.png)
![image](https://user-images.githubusercontent.com/22273954/120605847-1e1d6400-c481-11eb-98bd-bd0e24ee808c.png)

However, it differs in `BatchNorm` layers. For example,

![image](https://user-images.githubusercontent.com/22273954/120606319-a69c0480-c481-11eb-86b6-33a5a93053dd.png)

Since various `aux_weight` setting yields no difference in `conv` gradients, the optimization of the model during training should be exactly the same at every step. Is it meaningless to set different `aux_weight` or doing any operation after the `SoftmaxOutput` given the above finding? But why it differs in terms of `BatchNorm`?

I am quite confused and ask for help here. Thanks very much! 

This issue may be similar to [issue#1665](https://github.com/dmlc/gluon-cv/issues/1665).

**Comment by BebDong** [comment]
When using `softmax` or `log_softmax`, the gradients seem different with various `aux_weight`.

**Comment by BebDong** [comment]
@zhreshold Can you help?

**Comment by zhreshold** [comment]
The question is more appropriate for mxnet, but I can give you some hint here. The softmaxoutput operator was used to terminate the gradient path, where the operator expects the input(prior to softmax op) and label(the ground truth), so no gradient is passing through this operator. Hope this answer's your question.
